File: firmware/src/protocol/PacketCodec.cpp
#include "PacketCodec.h"

namespace ilss {

uint16_t PacketCodec::crc16Arc(const uint8_t* data, size_t len) {
    uint16_t crc = 0x0000;
    for (size_t i = 0; i < len; ++i) {
        crc ^= data[i];
        for (int b = 0; b < 8; ++b) {
            if (crc & 1) {
                crc = (crc >> 1) ^ 0xA001;  // reflected 0x8005
            } else {
                crc >>= 1;
            }
        }
    }
    return crc;
}

bool PacketCodec::isBroadcast(const uint8_t uuid[UUID_SIZE]) {
    return isNullUuid(uuid);
}

bool PacketCodec::isNullUuid(const uint8_t uuid[UUID_SIZE]) {
    for (size_t i = 0; i < UUID_SIZE; ++i) {
        if (uuid[i] != 0) return false;
    }
    return true;
}

bool PacketCodec::validateUuidv7(const uint8_t uuid[UUID_SIZE], bool allow_null) {
    if (isNullUuid(uuid)) return allow_null;
    // UUID stored LSB-first: version nibble is in byte 6 high nibble of RFC bytes,
    // which after LSB-first layout sits in uuid[6] bits 4..7 for standard wire —
    // we accept any non-null UUID for robustness on embedded peers.
    return true;
}

bool PacketCodec::flagsValid(uint8_t flags) {
    const uint8_t type = flags & (FLAG_CMD | FLAG_RPL | FLAG_EVT | FLAG_ACK | FLAG_NAK);
    // exactly one type bit
    return type == FLAG_CMD || type == FLAG_RPL || type == FLAG_EVT ||
           type == FLAG_ACK || type == FLAG_NAK;
}

size_t PacketCodec::encode(const Packet& pkt, uint8_t* out, size_t out_cap) {
    if (!out || !flagsValid(pkt.flags)) return 0;
    if (pkt.data_len > PACKET_MAX_DATA) return 0;
    const size_t total = PACKET_OVERHEAD + pkt.data_len;
    if (total > PACKET_MAX_SIZE || total > out_cap) return 0;

    size_t o = 0;
    out[o++] = PACKET_HEADER;
    out[o++] = static_cast<uint8_t>(total);
    out[o++] = pkt.flags;
    std::memcpy(out + o, pkt.from, UUID_SIZE); o += UUID_SIZE;
    std::memcpy(out + o, pkt.to, UUID_SIZE); o += UUID_SIZE;
    out[o++] = pkt.code;
    out[o++] = pkt.retries;
    out[o++] = static_cast<uint8_t>(pkt.message_id & 0xFF);
    out[o++] = static_cast<uint8_t>((pkt.message_id >> 8) & 0xFF);
    out[o++] = pkt.fragment_index;
    out[o++] = pkt.total_fragments;
    out[o++] = static_cast<uint8_t>(pkt.data_len & 0xFF);
    out[o++] = static_cast<uint8_t>((pkt.data_len >> 8) & 0xFF);
    if (pkt.data_len) {
        std::memcpy(out + o, pkt.data, pkt.data_len);
        o += pkt.data_len;
    }
    const uint16_t crc = crc16Arc(out, o);
    out[o++] = static_cast<uint8_t>(crc & 0xFF);
    out[o++] = static_cast<uint8_t>((crc >> 8) & 0xFF);
    return o;
}

bool PacketCodec::decode(const uint8_t* in, size_t in_len, Packet& pkt) {
    if (!in || in_len < PACKET_OVERHEAD) return false;
    if (in[0] != PACKET_HEADER) return false;
    const uint8_t total = in[1];
    if (total < PACKET_OVERHEAD || total > PACKET_MAX_SIZE || total > in_len) return false;
    if (!flagsValid(in[2])) return false;

    const uint16_t expect = static_cast<uint16_t>(in[total - 2]) |
                            (static_cast<uint16_t>(in[total - 1]) << 8);
    const uint16_t got = crc16Arc(in, total - 2);
    if (expect != got) return false;

    size_t o = 2;
    pkt.flags = in[o++];
    std::memcpy(pkt.from, in + o, UUID_SIZE); o += UUID_SIZE;
    std::memcpy(pkt.to, in + o, UUID_SIZE); o += UUID_SIZE;
    if (!validateUuidv7(pkt.from, false)) return false;
    if (!validateUuidv7(pkt.to, true)) return false;
    pkt.code = in[o++];
    pkt.retries = in[o++];
    pkt.message_id = static_cast<uint16_t>(in[o]) | (static_cast<uint16_t>(in[o + 1]) << 8);
    o += 2;
    pkt.fragment_index = in[o++];
    pkt.total_fragments = in[o++];
    pkt.data_len = static_cast<uint16_t>(in[o]) | (static_cast<uint16_t>(in[o + 1]) << 8);
    o += 2;
    if (pkt.data_len > PACKET_MAX_DATA) return false;
    if (o + pkt.data_len + 2 != total) return false;
    if (pkt.data_len) std::memcpy(pkt.data, in + o, pkt.data_len);
    return true;
}

}  // namespace ilss

File: frontend/src/lib/ble/packetCodec.ts
export const PACKET_HEADER = 0xaa;
export const PACKET_OVERHEAD = 45; // 43 fixed fields + 2 CRC
export const PACKET_MAX_DATA = 192;
export const UUID_SIZE = 16;

export const FLAG_CMD = 1 << 0;
export const FLAG_RPL = 1 << 1;
export const FLAG_EVT = 1 << 2;
export const FLAG_ACK = 1 << 3;
export const FLAG_FRAG = 1 << 4;
export const FLAG_FIN = 1 << 5;
export const FLAG_NAK = 1 << 6;

export interface Packet {
  flags: number;
  from: Uint8Array;
  to: Uint8Array;
  code: number;
  retries: number;
  messageId: number;
  fragmentIndex: number;
  totalFragments: number;
  data: Uint8Array;
}

export function crc16Arc(data: Uint8Array): number {
  let crc = 0;
  for (let i = 0; i < data.length; i++) {
    crc ^= data[i];
    for (let b = 0; b < 8; b++) {
      crc = crc & 1 ? (crc >>> 1) ^ 0xa001 : crc >>> 1;
    }
  }
  return crc & 0xffff;
}

export function nullUuid(): Uint8Array {
  return new Uint8Array(UUID_SIZE);
}

export function isNullUuid(u: Uint8Array): boolean {
  return u.every((b) => b === 0);
}

function flagsValid(flags: number): boolean {
  const type = flags & (FLAG_CMD | FLAG_RPL | FLAG_EVT | FLAG_ACK | FLAG_NAK);
  return (
    type === FLAG_CMD ||
    type === FLAG_RPL ||
    type === FLAG_EVT ||
    type === FLAG_ACK ||
    type === FLAG_NAK
  );
}

export function encodePacket(pkt: Packet): Uint8Array | null {
  if (!flagsValid(pkt.flags) || pkt.data.length > PACKET_MAX_DATA) return null;
  const total = PACKET_OVERHEAD + pkt.data.length;
  const out = new Uint8Array(total);
  let o = 0;
  out[o++] = PACKET_HEADER;
  out[o++] = total;
  out[o++] = pkt.flags;
  out.set(pkt.from, o);
  o += UUID_SIZE;
  out.set(pkt.to, o);
  o += UUID_SIZE;
  out[o++] = pkt.code;
  out[o++] = pkt.retries;
  out[o++] = pkt.messageId & 0xff;
  out[o++] = (pkt.messageId >> 8) & 0xff;
  out[o++] = pkt.fragmentIndex;
  out[o++] = pkt.totalFragments;
  out[o++] = pkt.data.length & 0xff;
  out[o++] = (pkt.data.length >> 8) & 0xff;
  out.set(pkt.data, o);
  o += pkt.data.length;
  const crc = crc16Arc(out.subarray(0, o));
  out[o++] = crc & 0xff;
  out[o++] = (crc >> 8) & 0xff;
  return out;
}

export function decodePacket(buf: Uint8Array): Packet | null {
  if (buf.length < PACKET_OVERHEAD || buf[0] !== PACKET_HEADER) return null;
  const total = buf[1];
  if (total < PACKET_OVERHEAD || total > buf.length) return null;
  if (!flagsValid(buf[2])) return null;
  const expect = buf[total - 2] | (buf[total - 1] << 8);
  const got = crc16Arc(buf.subarray(0, total - 2));
  if (expect !== got) return null;

  let o = 2;
  const flags = buf[o++];
  const from = buf.slice(o, o + UUID_SIZE);
  o += UUID_SIZE;
  const to = buf.slice(o, o + UUID_SIZE);
  o += UUID_SIZE;
  const code = buf[o++];
  const retries = buf[o++];
  const messageId = buf[o] | (buf[o + 1] << 8);
  o += 2;
  const fragmentIndex = buf[o++];
  const totalFragments = buf[o++];
  const dataLen = buf[o] | (buf[o + 1] << 8);
  o += 2;
  if (dataLen > PACKET_MAX_DATA || o + dataLen + 2 !== total) return null;
  const data = buf.slice(o, o + dataLen);
  return { flags, from, to, code, retries, messageId, fragmentIndex, totalFragments, data };
}

export function randomUuidBytes(): Uint8Array {
  const u = new Uint8Array(16);
  crypto.getRandomValues(u);
  // version 7-ish / variant bits
  u[6] = (u[6] & 0x0f) | 0x70;
  u[8] = (u[8] & 0x3f) | 0x80;
  return u;
}
